Improve section tag constants in archives

Brett208 opened the pull request:
See commit messages for details. This should close out issue #58. 

Tested by listing contents of a volume, and packing a new clm file.

DanRStevens commented on Archives/VolFile.cpp:
```diff
@@ -378,49 +387,37 @@ namespace Archives
-
-		// Check for the tag (MSB set)
-		if ((length & 0x80000000) != 0x80000000) {
+		if (tag.padding == VolPadding::TwoByte) {
```
This is wonderful syntax

Brett208 replied:
Thanks, yeah I think it worked well.

DanRStevens commented on Archives/VolFile.cpp:
```diff
@@ -506,4 +503,8 @@ namespace Archives
+
+	VolFile::SectionHeader::SectionHeader() {}
+	VolFile::SectionHeader::SectionHeader(std::array<char, 4> tag, uint32_t length, VolPadding padding) 
+		: tag(tag), length(length), padding(padding) {}
```
Good, though I'm not sure if this needs to be explicit. Public struct fields can be initialized with brace syntax. There may be some benefits to using constructor notation though. Maybe experiment with using `= default`, at least for the empty constructor.

DanRStevens commented on Archives/VolFile.h:
```diff
@@ -53,11 +53,20 @@ namespace Archives
+		enum class VolPadding : uint32_t
+		{
+			TwoByte = 0,
+			FourByte = 1 // Default value
```
Rather than a comment, perhaps this can be written as:
```
enum class VolPadding : uint32_t
{
	TwoByte = 0,
	FourByte = 1,
	Default = FourByte
}
```

Brett208 replied:
I didn't actually know you could use default like this on a c++ enum. I think that it might be a good idea to put the default in so it is clear to maybe a less astute reader who maybe doesn't notice/understand the optional argument syntax in the constructor.

DanRStevens replied:
If the constructor then sets the default value using `Default`, it makes the constructor less self-documenting.

If another class is written, where someone wanted to default to 2-byte padding, it would be weird to reuse the existing enum.

DanRStevens commented on Archives/VolFile.h:
```diff
@@ -53,11 +53,20 @@ namespace Archives
 			std::array<char, 4> tag;
 			uint32_t length : 31;
-			uint32_t padding : 1; // 0 = 4-byte boundary padding, 1 = 2-byte boundary padding
+			VolPadding padding : 1;
```
You may want to verify `sizeof(SectionHeader)` here. I believe in some cases, changing the type can break apart bit fields into separate storage locations. In particular, `uint32_t` is not `VolPadding`. It is close though, in that it's an `enum` based on `uint32_t`, so it might still work.

I'd recommend adding a `static_assert` check to ensure the size of the struct is and always remains the size required by the file format:
```
static_assert(sizeof(SectionHeader) == 8, "SectionHeader not of required size");
```

Brett208 replied:
This aspect of bit fields is certainly beyond my knowledge of the language. I placed the static_assert in the constructor of VolFile. It didn't give me any errors, so I guess on MSVC it is okay. 

Let me know if it is more appropriate to place it somewhere else though. I've never actually used static_assert before.

DanRStevens replied:
I would place it right after the struct definition. It is a compile time check, so it doesn't need to be placed within a function.

DanRStevens commented on Archives/VolFile.h:
```diff
@@ -79,7 +88,8 @@ namespace Archives
 		uint32_t ReadTag(const std::array<char, 4>& tagName);
-		void WriteTag(StreamWriter& volWriter, uint32_t length, const char *tagText);
+		//void WriteTag(StreamWriter& volWriter, uint32_t length, const char *tagText);
+		void VolFile::WriteTag(StreamWriter& volWriter, uint32_t length, const std::array<char, 4>& tagName);
```
You have a `VolFile::` prefix here.

Should also remove the old commented out code so it doesn't confuse people in the future.

Or maybe consider removing this function entirely, and just passing a constructed `StreamHeader` directly to `volWriter.Write`. It should be possible to even pass an unnamed temporary value. Though that may lose out on the nice exception message you provided in the function.

Brett208 replied:
On further review, I think this function should be removed entirely. The error reporting currently in write tag will only catch problems with the actual write process that are not specific to the tag. It would just catch if there was a generic write problem that happened to occur during the write process, if that makes sense. 

I should be better about deleting the class prefix in the header files. I tend to create the function first, then paste the prototype into the header file and get sloppy about cleaning it up.

DanRStevens commented on Archives/VolFile.cpp:
```diff
@@ -9,6 +9,14 @@
+	const std::array<char, 4> TagVOLH{ 'v', 'o', 'l', 'h' }; // Header tag
+	const std::array<char, 4> TagVOLS{ 'v', 'o', 'l', 's' }; // Filename table tag
+	const std::array<char, 4> TagVOLI{ 'v', 'o', 'l', 'i' }; // Index table tag
+	const std::array<char, 4> TagVBLK{ 'V', 'B', 'L', 'K' }; // Packed file tag
```
I can live with this. The tags are being initialized by the actual string data. Uses are name checked so there are no typos.

I wonder if there might be a way to improve the intialization syntax using the StackOverflow answer from [How to initializez a `std::array<char, n>` with a string literal, omitting the trailing zero](https://stackoverflow.com/questions/33484233/how-to-initialize-a-stdarraychar-n-with-a-string-literal-omitting-the-trail). Admittedly I don't entirely understand the answer, though it looks short enough to study and figure out.

Brett208 replied:
I have split feelings about this. I wish there was a built in language solution to this problem. For these short arrays, it isn't a big deal though it would be nice to improve the syntax.

DanRStevens replied:
Let's leave this for another day then. It may involve some learning and experimentation to come up with a satisfactory solution. I don't want to hold up the other changes on this.

DanRStevens commented on Archives/WaveFile.h:
```diff
@@ -1,13 +1,14 @@
+	const std::array<char, 4> tagRIFF{ 'R', 'I', 'F', 'F' };
+	const std::array<char, 4> tagWAVE{ 'W', 'A', 'V', 'E' };
+	const std::array<char, 4> tagFMT_{ 'f', 'm', 't', ' ' };
+	const std::array<char, 4> tagDATA{ 'd', 'a', 't', 'a' };
```
This is a little easier to verify for correctness. I like that.

DanRStevens commented on Archives/VolFile.cpp:
```diff
@@ -264,12 +272,13 @@ namespace Archives
 
-		WriteTag(volWriter, 0, "volh");
+		WriteTag(volWriter, 0, TagVOLH);
+		//WriteTag(volWriter, 0, "volh");
```
Let's clear out this old line of code.

Brett208 replied:
Sounds good.

Brett208 commented on Archives/VolFile.cpp:
```diff
@@ -378,49 +387,37 @@ namespace Archives
 
 	// Writes a section tag to the open output file.
-	void VolFile::WriteTag(StreamWriter& volWriter, uint32_t length, const char *tagText)
+	void VolFile::WriteTag(StreamWriter& volWriter, uint32_t length, const std::array<char, 4>& tagName)
```
Agreed, I started not passing by reference later in the changes but never made it back to pull the reference out here. Thanks for catching.

Brett208 commented on Archives/VolFile.cpp:
```diff
@@ -378,49 +387,37 @@ namespace Archives
-		buffer[0] = *(int*)tagText;
-		// Set padding bit to indicate 4 byte padding
-		buffer[1] = length | 0x80000000;
+		SectionHeader sectionHeader(tagName, length);
```
I set the constructor so if the padding value is not added then it will automatically be set to 1. 

What do you think of this format, where an argument has a default value? Maybe it isn't appropriate here?

DanRStevens replied:
You're right. I didn't see that until after I submitted the review. I was focused on the implementation of the constructor, where the default value isn't listed.

The code is fine as is.

DanRStevens commented on Archives/VolFile.h:
```diff
@@ -57,7 +57,8 @@ namespace Archives
 			TwoByte = 0,
-			FourByte = 1 // Default value
+			FourByte = 1,
+			default = FourByte
```
Err, `default` is a reserved word. This might cause problems. The syntax highlighting on GitHub kind of points that out here.

DanRStevens commented on Archives/VolFile.cpp:
```diff
@@ -19,6 +19,8 @@ namespace Archives
 
 	VolFile::VolFile(const char *fileName) : ArchiveFile(fileName), archiveFileReader(fileName)
 	{
+		static_assert(sizeof(SectionHeader) == 8, "SectionHeader not of required size");
```
This should be placed right after the `struct SectionHeader` definition. It doesn't need to be in a function, so it makes sense to put it right after the struct it applies to.

Brett208 commented:
Okay, I thought it involved using the default keyword in enum, not defining a new member that was Default, but not the keyword. I deleted it. Also moved the static_assert to the header file.
